### PR TITLE
Improve dataset

### PR DESCRIPTION
**Pull request by zealseeker**
* keep the graphs of a chunk in memory to make it faster to traverse

The original framework is too slow because it read a file each time `__getitem__` was called. Since at most time, the data set is traversed rather than being picked, I think we can add the graphs into cache.

Experiment:
For `chunk_size=1000`, it took 192s to read 320 (10 batches) compounds. In this revised version, it took only 66s to traverse the whole training set.

```python
from dataset import PCBADataset, split_dataset
from torch.utils.data import DataLoader
from utils import collate_molgraphs
from time import time

batch_size = 32
chunk_size = 1000

print('Preprocessing....')
crt_time = time()
dataset = PCBADataset(chunk_size=chunk_size)
trainset, valset, testset = split_dataset(dataset)
train_loader = DataLoader(trainset, batch_size=batch_size, collate_fn=collate_molgraphs)
val_loader = DataLoader(valset, batch_size=batch_size, collate_fn=collate_molgraphs)
test_loader = DataLoader(testset, batch_size=batch_size, collate_fn=collate_molgraphs)
print('Data loaded, time: %.3f' % (time()-crt_time))

crt_time = time()
for batch_id, batch_data in enumerate(train_loader):
    pass

print('Traversed the training set, time: %.3f' % (time()-crt_time))
```

### CODE REVIEW

**Comment by mufeili** [comment]
Thanks Hongbin for the suggestion.

Have you tried using a `chunk_size=1`? The reason I used `chunk_size=1` and loaded the preprocessed graphs every time is because we typically shuffle the dataset at the beginning of each epoch and it's highly likely that two molecules fetched come from different pickled objects. In that case, using a `chunk_size=1` and loading the graph every time can be a lot more faster.

**Comment by zealseeker** [comment]
Setting `chunk_size=1` can be a solution, which is much faster. But it is a bit weird...

As for data splitting,  what we always do is to shuffle before the data are "dumped" so that when training, we can read the files and yield each item by a generator.

Another concern is that when the data set is very large (1 million), file system might be slower if we make every item a file. I am not sure about it.

**Comment by mufeili** [comment]
The experience from the ML community is that shuffling the dataset at the beginning of each epoch helps making the model more robust and avoids weird optimization trajectories, but I agree this might much depend on the particular case.

We are developing a data format for serializing DGLGraphs in https://github.com/dmlc/dgl/pull/728 so that we can store all graphs in one file. It will be much faster than `pickle` as we will not need to load all graphs at once but only the ones that are needed. Before that function gets completed, I'm proposing pickling one DGLGraph at a time as a temporary workaround.

**Comment by zealseeker** [comment]
OK. Expecting your new serializing method. It's important to support training large data sets.
I am closing the PR and may reopen it if any better solution is found.
